spring-kafka/src/main/java/org/springframework/kafka/listener/KafkaMessageListenerContainer.java

```diff
@@ -443,89 +480,148 @@ public void run() {
 			}
 		}
 
-		private void invokeListener(final ConsumerRecord<K, V> record) {
-			try {
-				if (this.acknowledgingMessageListener != null) {
-					this.acknowledgingMessageListener.onMessage(record, new Acknowledgment() { //NOSONAR - long anon.
-
-						@Override
-						public void acknowledge() {
-							if (ListenerConsumer.this.ackMode.equals(AckMode.MANUAL)) {
-								updateManualOffset(record);
-							}
-							else if (ListenerConsumer.this.ackMode.equals(AckMode.MANUAL_IMMEDIATE)
-									|| ListenerConsumer.this.ackMode.equals(AckMode.MANUAL_IMMEDIATE_SYNC)) {
-								ackImmediate(record);
-							}
-							else {
-								throw new IllegalStateException("AckMode must be MANUAL, "
-										+ "MANUAL_IMMEDIATE, or MANUAL_IMMEDIATE_SYNC for manual acks");
-							}
-						}
-
-						private void ackImmediate(final ConsumerRecord<K, V> record) {
-							if (Thread.currentThread().equals(ListenerConsumer.this.consumerThread)) {
-								Map<TopicPartition, OffsetAndMetadata> commits = Collections.singletonMap(
-										new TopicPartition(record.topic(), record.partition()),
-										new OffsetAndMetadata(record.offset() + 1));
-								if (ListenerConsumer.this.logger.isDebugEnabled()) {
-									ListenerConsumer.this.logger.debug("Committing: " + commits);
-								}
-								if (ListenerConsumer.this.ackMode.equals(AckMode.MANUAL_IMMEDIATE)) {
-									ListenerConsumer.this.consumer.commitAsync(commits,
-											ListenerConsumer.this.commitCallback);
-								}
-								else {
-									ListenerConsumer.this.consumer.commitSync(commits);
-								}
-							}
-							else {
-								throw new IllegalStateException(
-										"With " + ListenerConsumer.this.ackMode.name()
-										+ " ack mode, acknowledge() must be invoked on the consumer thread");
-							}
-						}
+		private ConsumerRecords<K, V> checkPause(ConsumerRecords<K, V> unsent) {
+			if (this.paused && this.recordsToProcess.size() < this.queueDepth) {
+				// Listener has caught up.
+				this.consumer.resume(
+						this.assignedPartitions.toArray(new TopicPartition[this.assignedPartitions.size()]));
+				this.paused = false;
+				if (unsent != null) {
+					try {
+						sendToListener(unsent);
+					}
+					catch (InterruptedException e) {
+						Thread.currentThread().interrupt();
+						throw new KafkaException("Interrupted while sending to listener", e);
+					}
+				}
+				return null;
+			}
+			return unsent;
+		}
 
-						@Override
-						public String toString() {
-							return "Acknowledgment for " + record;
-						}
+		private boolean sendToListener(final ConsumerRecords<K, V> records) throws InterruptedException {
+			if (this.pauseEnabled) {
+				return !this.recordsToProcess.offer(records, this.pauseAfter, TimeUnit.MILLISECONDS);
+			}
+			else {
+				this.recordsToProcess.put(records);
```

Comment:
According `put()` JavaDocs:
```
* Inserts the specified element into this queue, waiting if necessary
* for space to become available.
```
So, if `pauseEnabled == false` we are blocked here in case of slow listener and therefore lose our `Consumer` because there is no heartbeat poll in the `run()` above.
Isn't it ?